Gray vs multigroup

brryan opened the pull request:
## Background

We want to have both gray and multigroup modes

One question is whether this should be a continuous energy or a multigroup representation. `Singularity-opac` right now always accepts a continuous frequency argument; maybe this needs to be expanded upon with a multigroup integer index variant. Should multigroup support nonuniformly binned data?

## Description of Changes

- [x] Switch to @RyanWollaeger's [Singularity-opac PR 59](https://github.com/lanl/singularity-opac/pull/59)
- [x] Split `jaybenne` package initialization into gray vs mg depending on which singularity variants are present
- [x] Support both mean and frequency-dependent opacity variants in configure system
- [x] Template sourcing on gray vs mg
- [x] Template transport on gray vs mg

## Checklist

- [ ] ~New features are documented~
- [x] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files
- [x] [Singularity-opac PR 59](https://github.com/lanl/singularity-opac/pull/59) is merged

brryan committed:
ryan's mean singularity-opac

brryan committed:
Multiple initialize commands

brryan committed:
fix some issues

brryan committed:
Sample MG CDF

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,37 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                    emis += vmesh(b, fj::emission_cdf(n), k, j, i);
+                    // Normalize emission CDF
+                    vmesh(b, fj::emission_cdf(n), k, j, i) /=
+                        vmesh(b, fj::emission_cdf(n_nubins - 1), k, j, i);
```
```suggestion
                        vmesh(b, fj::emission_cdf(n_nubinsd - 1), k, j, i);
```
? Also, should the normalization occur after the aggregation?

brryan replied:
Yes good catch with the `n_nubinsd`. 

I think the normalization is properly done after aggregation because the previous loop is like `cdf(n) = emis + cdf(n-1)` right? Maybe I'm missing something

RyanWollaeger replied:
I couldn't see where `emis` was getting used to set the `fj::emission_cdf` values in this or the previous loops... I could certainly be missing something.

RyanWollaeger replied:
Wait - nevermind - I see where `emission_cdf` is getting set above.

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,37 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                  }
+                  for (int n = 0; n < n_nubinsd; n++) {
+                    // Get total emissivity
+                    emis += vmesh(b, fj::emission_cdf(n), k, j, i);
```
It looks like `emis` is set to -1 (poisoned) at line 113 above, but then incremented here without being re-initialized to 0?

brryan replied:
Nice catch, thanks! (fixed)

brryan committed:
n_nubinsd, emis poison fix, work on transporters

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,38 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                  for (int n = 1; n < n_nubinsd; n++) {
+                    const Real nu = std::exp(std::log(numind) + (n + 0.5) * dlnu);
+                    vmesh(b, fj::emission_cdf(n), k, j, i) =
+                        opac->EmissivityPerNu(rho, temp, nu) +
```
Quick question on the units here - from the `singularity-opac` README, it looks like `EmissivityPerNu` returns units as energy density per time per frequency. Should there be a `dnu` multiplying `opac->EmissivityPerNu(rho, temp, nu)` here?

RyanWollaeger replied:
Think this is resolved.

brryan committed:
Fixing little errors

brryan committed:
Need to fix UpdateDerivedTransportFields

brryan committed:
multiply Jnu by dnu

brryan committed:
fix cdf integration

brryan committed:
inf example cycles

brryan committed:
Modified inputs... but stepdiff is nan...

brryan committed:
OK stepdiff runs now but the test fails

RyanWollaeger commented:
@brryan - sorry to lose track of this - are there any significant further updates pending?

RyanWollaeger commented on src/jaybenne/transport.cpp:
```diff
@@ -111,20 +127,27 @@ TaskStatus TransportPhotons(MeshData<Real> *md, const Real t_start, const Real d
+            Real aa;
+            if constexpr (FT == FrequencyType::gray) {
+              ss = mscattering.RosselandMeanTotalScatteringCoefficient(rho, temp);
+              aa = mopacity.RosselandMeanAbsorptionCoefficient(rho, temp);
```
Could this absorption use the option Rosseland option in `AbsorptionCoefficient`? Actually, should the mean scattering class have had an update to toggle between Rosseland and Planck means (sorry to have missed this if so)?

brryan replied:
Yeah youre right the general absorption option would be better here. Should the Rosseland and Planck named versions of this function just be deleted in singularity-opac?

RyanWollaeger replied:
That is a good question - thought they may want to be kept for codes where Planck and Rosseland opacities are necessarily used differently (e.g. diffusion/P1), for more self-descriptiveness. However, if the enum is used then it should be pretty clear what the opacity type is.

RyanWollaeger replied:
Will make the change here, at any rate, and maybe update `singularity-opac` to have the corresponding `ScatteringCoefficient(rho, temp, mode)` function.

RyanWollaeger commented:
@brryan - if you are still catching updates here, is this ready to merge?

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -31,8 +31,21 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+    n_nubins = jb_pkg->template Param<int>("n_nubins");
+    numin = jb_pkg->template Param<Real>("numin");
+    numax = jb_pkg->template Param<Real>("numax");
+  }
```
Maybe just a comment for posterity, but since `n_nubins`, `numin`, and `numax` are "poisoned" to invalid values above, should there be assertions for valid values after this block?

brryan replied:
I'm torn on this point. On the one hand, assertion checks like you say would be extra safety, but I think code can get cluttered when it becomes mostly DBC checks. And the poisoned values should be pretty obvious down the line... What the right balance of those two considerations is for the jaybenne project is for you and others to decide :)

RyanWollaeger replied:
Great point. This seems like a sensible enough argument to not add them to me.

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,44 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                } else if constexpr (FT == FrequencyType::multigroup) {
+                  // Construct emission CDF
+                  const Real dlnu = (std::log(numaxd) - std::log(numind)) / n_nubinsd;
+                  Real nu = std::exp(std::log(numind) + 0.5 * dlnu);
```
```suggestion
                  // this is the mid-point of the 1st group in log-space: nu=exp(log(numin) + 0.5*dlnu)
                  Real nu = numind * std::exp(0.5 * dlnu);
```
Since the log and exponent cancel (just a suggestions)?

brryan replied:
Oh right this does save a log, good thinking. Might be good to add a comment as well explaining what the expression "really" is though

RyanWollaeger replied:
A log saved is a log earned

RyanWollaeger replied:
Will do on the comment (going to pull this branch down to add a test).

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,44 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                  Real dnu = nu * dlnu;
+                  vmesh(b, fj::emission_cdf(0), k, j, i) =
+                      opac.EmissivityPerNu(rho, temp,
+                                           std::exp(std::log(numind) + 0.5 * dlnu)) *
```
```suggestion
                                           numind * std::exp(0.5 * dlnu)) *
```
Same suggestion...

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,44 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                                           std::exp(std::log(numind) + 0.5 * dlnu)) *
+                      dnu;
+                  for (int n = 1; n < n_nubinsd; n++) {
+                    nu = std::exp(std::log(numind) + (n + 0.5) * dlnu);
```
```suggestion
                    // this is the mid-point of group n in log-space: nu=exp(log(numin)+(n+0.5)*dlnu)
                    nu = numind * std::exp((n + 0.5) * dlnu);
```

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,44 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                  // Construct emission CDF
+                  const Real dlnu = (std::log(numaxd) - std::log(numind)) / n_nubinsd;
+                  Real nu = std::exp(std::log(numind) + 0.5 * dlnu);
+                  Real dnu = nu * dlnu;
```
Sorry if missing something here - does `dlnu` need to be exponentiated here to get the frequency group width?, E.g. something like:
```suggestion
                  Real dnu = nu * (std::exp(dlnu) - 1.0);
```

brryan replied:
I think we're allowed to use `dx = x d log x`

RyanWollaeger replied:
Ah - right - good point. However, these formulae only agree at small `dlnu` - for coarse group widths would there be an issue covering the intended range of frequencies?  Then again, the absolute error between these is only O(`dlnu^3`) when the midpoint `nu` is used in your formula, as is true here, with a relative error of O(`dlnu^2`). This implies the relative discrepancy between group widths for these formulae should only be 1% when 10 groups are used, which is maybe not worth worrying about... sorry was just curious - thanks!

RyanWollaeger commented on src/jaybenne/sourcing.cpp:
```diff
@@ -87,12 +101,44 @@ TaskStatus SourcePhotons(T *md, const Real t_start, const Real dt) {
+                      dnu;
+                  for (int n = 1; n < n_nubinsd; n++) {
+                    nu = std::exp(std::log(numind) + (n + 0.5) * dlnu);
+                    dnu = dlnu * nu;
```
Same question here - should note the previous suggestion is only exactly correct if `nu` is the lower-bound, but it looks like the logarithmic midpoint, rather.

brryan commented:
> @brryan - if you are still catching updates here, is this ready to merge?

Hey @RyanWollaeger sorry for my delay, yes I think this is ready to merge (or at least I don't remember any outstanding issues). It's not really tested so it might be useful to add a test of actual working multigroup, like frequency-dependent bremss thermalization that shows the gas temperature undershoot the equilibrium at first.

brryan committed:
Merge branch 'develop' into brryan/gray_vs_mg

RyanWollaeger committed:
Apply suggestions from code review

+ Cancel log and exponential where possible.
+ Add comment explaining nu formula is log-space group midpoint.

RyanWollaeger committed:
Make a few minor updates.

+ Fix the formatting using style/format.sh.
+ Use new AbsorptionCoefficient API for mean opacity.
+ Update some parts of the README.md file.

Note: similar to MeanOpacity's Emissivity, AbsorptionCoefficient
defaults to Rosseland-weighted grey opacity.

RyanWollaeger commented:
Looks like `stepdiff` is failing with the current changeset - that should be unaffected unless there is an issue switching to the grey API (or one of the API functions).

RyanWollaeger committed:
Cancel more exp(log()) occurrences.

RyanWollaeger commented:
@brryan or @pdmullen - I should know this, but how do we run the full test suite locally (i.e. not with CI)?

RyanWollaeger commented:
> @brryan or @pdmullen - I should know this, but how do we run the full test suite locally (i.e. not with CI)?

N/m - it occurred to me I can just look at how the CI is doing it... yikes.

RyanWollaeger committed:
Do not divide particle energy weight by frequency.

+ Remove division by frequency in source particle creation.
+ Update singularity-opac commit to correct mean opacity accesses.
+ Have regressions fail if mean errors are NaN.

RyanWollaeger commented:
@pdmullen - the most recent commit _should_ get the existing tests to pass. Still need to add an actual multi-frequency test.

RyanWollaeger commented:
> @pdmullen - the most recent commit _should_ get the existing tests to pass. Still need to add an actual multi-frequency test.

Guess not - the passed for me locally, at least.

RyanWollaeger committed:
+ Update singularity-eos submodule for spiner (databox) update.
+ Update parthenon submodule.

Note: the spiner header is included through singularity-eos, not
singularity-opac.

RyanWollaeger commented:
The `singularity-eos` `spiner` submodule update is not yet merged to the main repo - just seeing what happens when the updated commit hash is provided from a PR issued from a forked repo.

RyanWollaeger commented:
> The `singularity-eos` `spiner` submodule update is not yet merged to the main repo - just seeing what happens when the updated commit hash is provided from a PR issued from a forked repo.

It breaks, as reasonably expected.

RyanWollaeger committed:
Include generated headers and add_subdirectory for singularity-eos.
(Thanks Adam Dempsey.)

RyanWollaeger committed:
Attempt to fix grey DDMC.

+ Use mean opacities in transport_ddmc.
+ Add PARTHENON_REQUIRE gray assertions for DDMC.

Commit:
Merge remote-tracking branch 'origin/develop' into brryan/gray_vs_mg

Commit:
Use null initialization about constexpr branching

Commit:
Fix scope issue in tasklist construction

Commit:
Update patch path

RyanWollaeger commented:
What counts as documentation (looking at the checklist in the description)?

pdmullen commented:
> What counts as documentation (looking at the checklist in the description)?

Let's have a dedicated PR to add documentation for `jaybenne`.  We are missing this entirely at the moment.

RyanWollaeger committed:
Add frequency redistribution from effective scattering to transport.cpp.

Note: this is the same as the emission CDF sampling, which fixes
particle frequencies at the log-center of the sampled group.

RyanWollaeger committed:
Add 16-group multigroup version of inf.in (infinite medium) input.

+ Use Bremsstrahlung multifrequency opacity option.
+ Use more physical initial temperature values in inf*.in inputs.

RyanWollaeger commented:
Added a 16 group infinite medium test that uses the Bremsstrahlung opacity, but haven't added it to the CI - it is pretty fast though, since the time step is small.  If there are no objections (e.g. making the test better, which is a reasonable request), will add it to the CI test suite...

Commit:
Bump parth and singularity-eos

Commit:
Bump singularity-opac
